System.Drawing netstandard1.6 compat: Update P/Invoke declarations

qmfrederik opened the pull request:
The P/Invoke declarations for GDI+ use `CharSet.Auto` and `UnmanagedType.CustomMarshaler`, which are not available on `netstandard1.6`.

To enable basic support for `netstandard1.6`, define these values as constants, keep the current values on non-`netstandard1.6` and use fallback values on `netstandard1.6`.

qmfrederik commented:
/cc @akoeplinger With all the changes that enabled netstandard2.0 support, basic netstandard1.6 support for System.Drawing is not that much of a stretch.

akoeplinger commented on mcs/class/System.Drawing/System.Drawing/gdipFunctions.cs:
```diff
@@ -979,22 +988,22 @@ static internal void CheckStatus (Status status)
 		// Image functions
-		[DllImport(GdiPlus, CharSet=CharSet.Auto)]
+		[DllImport(GdiPlus, CharSet=GdiPlusCharSet)]
 		internal static extern Status GdipLoadImageFromFile ( [MarshalAs(UnmanagedType.LPWStr)] string filename, out IntPtr image );
```
I wonder if we can't just use CharSet.Unicode in the Mono implementation as well here and in most other cases. It's marshalled as a Unicode/LPWStr anyway after all.

qmfrederik replied:
You're right:
- All strings have an explicit attribute on them forcing `LPWStr` marshaling, so it seems the `CharSet` attribute is redundant.
- Most (but not all) functions on the C side only expose a Unicode equivalent - e.g. [`GdipPrivateAddFontFile`](https://github.com/mono/libgdiplus/blob/7d30e92a6cfd6882cd57efeb7fb7b21b31ae2736/src/fontcollection.h#L41) only takes a `GDIPCONST WCHAR*` and there's no `GdipPrivateAddFontFileA`/`GdipPrivateAddFontFileW` variation, so ANSI would never be correct.
- I'm guessing Mono no longer supports Windows 95, 98 or Me so backwards Windows compatibility shouldn't be an issue, either

So there are two options:
- Remove the `CharSet` property from the `DllImport` attribute alltogether
- Using `Unicode` everywhere (since that's what libgdiplus expects)

Let me know which one you prefer and I'll update the PR.

akoeplinger replied:
I'd prefer making the CharSet Unicode explicitly where possible.

The cases where this isn't possible are where LOGFONT is involved, correct?

qmfrederik replied:
I checked, for `LOGFONT` there's a `LOGFONTA` and `LOGFONTW` both in [libgdiplus](https://github.com/mono/libgdiplus/blob/1cbd58ef57c208a8517a772f3c8416169d7dfd9b/src/win32structs.h#L151) and [Windows](https://msdn.microsoft.com/en-us/library/windows/desktop/dd145037(v=vs.85).aspx).

So we should be able to just update the` LOGFONT` struct to use Unicode and have all the P/Invoke function declarations use Unicode as well.

I'll make the change and push an update.

akoeplinger commented on mcs/class/System.Drawing/System.Drawing/gdipFunctions.cs:
```diff
@@ -47,6 +47,15 @@ namespace System.Drawing
+
+#if NETSTANDARD1_6
+		public const CharSet GdiPlusCharSet = CharSet.Unicode;
+		public const UnmanagedType GdiPlusCustomMarshaler = UnmanagedType.Interface;
```
does using UnmanagedType.Interface actually work in NS16/netcore or are those methods just never hit?

qmfrederik replied:
The custom marshaler is always `ComIStreamMarshaler`, and [there's a comment at the definition of `ComIStreamMarshaler`](https://github.com/mono/mono/blob/master/mcs/class/System.Drawing/System.Drawing/ComIStreamMarshaler.cs#L46) that basically reads:

> This custom marshaler should be replaced with UnmanagedType.Interface
> Provides identical behaviour under Mono and .NET Framework

That comment is 12 years old, though :)

akoeplinger replied:
ok, let's keep that code for now then :smile:

migueldeicaza commented:
We can not take code that switches Auto to Unicode, because Unicode on Mono/Unix means UTF16, and it breaks the bridge for us.

akoeplinger commented:
@migueldeicaza what do you mean with "it breaks the bridge for us" ?

As far as I can see all the libgdiplus code this is pinvoking into already takes a WCHAR which means UTF16.

qmfrederik commented:
@migueldeicaza @akoeplinger 

I think `gdipFunctions.cs` at this time is basically inconsistent, you see two patterns in the P/Invoke declarations for methods which take strings:
- `CharSet=CharSet.Unicode` which no `MarshalAs` attribute on the string arguments
- `CharSet=CharSet.Auto` with `[MarshalAs(UnmanagedType.LPWStr)]` on all string arguments.

I think that's a historical inconsistency and this PR addresses that, in the unmanaged world (`libgdiplus` and GDI+) these functions always have the same declaration - they use `GDIPCONST WCHAR *`.

Here are two examples:
```
[DllImport(GdiPlus, CharSet=CharSet.Unicode)]
static internal extern Status GdipDrawString (IntPtr graphics, string text, int len, IntPtr font, ref RectangleFrc, IntPtr format, IntPtr brush);
```

which maps to

```
GpStatus GdipDrawString (GpGraphics *graphics, GDIPCONST WCHAR *string, int length, GDIPCONST GpFont *font,  GDIPCONST RectF *layoutRect, GDIPCONST GpStringFormat *stringFormat, GpBrush *brush);
```

and

```
[DllImport(GdiPlus, CharSet=CharSet.Auto)]
internal static extern Status GdipLoadImageFromFile ( [MarshalAs(UnmanagedType.LPWStr)] string filename, out IntPtr image );
```

which maps to:

```
GpStatus WINGDIPAPI GdipLoadImageFromFile (GDIPCONST WCHAR *file, GpImage **image); 
```

I checked all the methods for which the P/Invoke signature changed from `CharSet.Auto` to `CharSet.Unicode`. They all take, or have `W`-variants which take, `LPWStr`.

Here's the complete list:
* `GdipLoadImageFromFile`, `GdipLoadImageFromFileICM`, `GdipCreateBitmapFromFile`, `GdipCreateBitmapFromFileICM`, `GdipSetImageAttributesOutputChannelColorProfile`, `GdipPrivateAddFontFile`, `GdipCreateFontFamilyFromName`, `GdipCreateMetafileFromFile`, `GdipGetMetafileHeaderFromFile`: 
  - Switched from `Auto` to `Unicode`.
  - Has one `string` parameter which is already marshaled as `LPWStr`. 

* `GdipGetLogFont`, `GdipCreateFontFromLogfont`, `CreateFontIndirect`: 
  - Take a `LOGFONT`. There's `LOGFONTA` and `LOGFONTW` in `libgdiplus` and GDI+ on Windows.
  - Converted `LogFont` in the managed side of things to use Unicode marshaling

Based on my knowledge, this should not introduce any breaking changes (and rather improvie consistency in `gdipFuncions.cs`) -- but happy to adapt as you see fit.

qmfrederik commented:
I ran the unit tests on this, and except for `LOGFONT` which I need to understand better, none of the changes to Unicode caused any of the tests to fail.

I updated the PR to exclude the `LOGFONT`-related changes and all `System.Drawing` tests fail.

@migueldeicaza Would you mind reviewing this one more time? Based on how I read the P/Invoke declarations (see my previous comment) the strings are already treated as UTF16 so there are no functional changes - all tests passing seems to confirm that.
